src/message/dkim.rs

```diff
@@ -0,0 +1,475 @@
+        domain: String,
+        private_key: DkimSigningKey,
+    ) -> DkimConfig {
+        DkimConfig {
+            selector,
+            domain,
+            private_key,
+            headers: vec![
+                "From".to_string(),
+                "Subject".to_string(),
+                "To".to_string(),
+                "Date".to_string(),
+            ],
+            canonicalization: DkimCanonicalization {
+                header: DkimCanonicalizationType::Simple,
+                body: DkimCanonicalizationType::Relaxed,
+            },
+        }
+    }
+    /// Set the signing key with given signing algorithm for a DkimConfig
+    pub fn set_signing_key(&mut self, private_key: DkimSigningKey) {
+        self.private_key = private_key;
+    }
+    /// Create a DkimConfig
+    pub fn new(
+        selector: String,
+        domain: String,
+        private_key: DkimSigningKey,
+        headers: Vec<String>,
+        canonicalization: DkimCanonicalization,
+    ) -> DkimConfig {
+        DkimConfig {
+            selector,
+            domain,
+            private_key,
+            headers,
+            canonicalization,
+        }
+    }
+}
+
+/// Create a Headers struct with a Dkim-Signature Header created from given parameters
+fn dkim_header_format(
+    config: &DkimConfig,
+    timestamp: String,
+    headers_list: String,
+    body_hash: String,
+    signature: String,
+) -> Headers {
+    let mut headers = Headers::new();
+    let header_name =
+        dkim_canonicalize_header_tag("DKIM-Signature".to_string(), config.canonicalization.header);
+    let header_name = HeaderName::new_from_ascii(header_name).unwrap();
+    headers.append_raw(header_name, format!("v=1; a={signing_algorithm}-sha256; d={domain}; s={selector}; c={canon}; q=dns/txt; t={timestamp}; h={headers_list}; bh={body_hash}; b={signature}",domain=config.domain, selector=config.selector,canon=config.canonicalization,timestamp=timestamp,headers_list=headers_list,body_hash=body_hash,signature=signature,signing_algorithm=config.private_key.get_signing_algorithm()));
+    headers
+}
+
+/// Canonicalize the body of an email
+fn dkim_canonicalize_body(body: &[u8], canonicalization: DkimCanonicalizationType) -> Vec<u8> {
+    static RE: Lazy<BRegex> = Lazy::new(|| BRegex::new("(\r\n)+$").unwrap());
+    static RE_DOUBLE_SPACE: Lazy<BRegex> = Lazy::new(|| BRegex::new("[\\t ]+").unwrap());
+    static RE_SPACE_EOL: Lazy<BRegex> = Lazy::new(|| BRegex::new("[\t ]\r\n").unwrap());
+    match canonicalization {
+        DkimCanonicalizationType::Simple => RE.replace(body, &b"\r\n"[..]).into_owned(),
+        DkimCanonicalizationType::Relaxed => {
+            let body = RE_DOUBLE_SPACE.replace_all(body, &b" "[..]).into_owned();
+            let body = RE_SPACE_EOL.replace_all(&body, &b"\r\n"[..]).into_owned();
+            RE.replace(&body, &b"\r\n"[..]).into_owned()
+        }
+    }
+}
+
+/// Canonicalize the value of an header
+fn dkim_canonicalize_header_value(
```

Comment:
I found an issue with a header that had a leading space. This was using a relaxed mode.

```
Subject:  Protecting your device information with Private Set Membership
```